### PR TITLE
Fix output of podman --remote top

### PR DESCRIPTION
**Pull request by rhatdan**
Fixes: https://github.com/containers/podman/issues/19176

#### Does this PR introduce a user-facing change?

```release-note
Corrected parsing of top ps options on podman-remote.
```

### CODE REVIEW

**Review by Luap99 — COMMENTED** [review]
This will juts move the same problem in the other direction and cause problem for space separated arguments.

We simply cannot send an array as single string to the server. We should instead change the server and client to send and accept an actual array. Only then we can keep the correct args.

**Comment by rhatdan** [comment]
That would be a breaking change. This fixes the existing issue, and matches the local output.

**Comment by rhatdan** [comment]
Take a look at the current patch, I got the API tests to work again, and now the passing of options works mostly the same.

**Comment by rhatdan** [comment]
-o "pid comm"

Looks like a cobra failure?

./bin/podman top dan -eo "pid comm"
DAN [dan -eo pid comm]

**Comment by rhatdan** [comment]
Never mind cobra is correct.

**Comment by Luap99** [comment]
No it's not a cobra problem, arrays are just printed like that. You can print each single element to see it. I think the issue is that shlex.Split() is used in Top() function. There seems to be a difference between what docker API wants and want we are doing. So I think this needs more investigation on what the expect behaviour should be when using these special formats.

**Comment by rhatdan** [comment]
Yes, @vrothberg The shlex code was your fix.  Do you have any idea what is going on. 
I believe my code is correct fixing this issue we are seeing, but podman top -eo "pid comm" is very broken because of the shlex.Split.

File: test/system/085-top.bats [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bats
+
+load helpers
+
+@test "podman top - basic tests" {
+    rand_name=$(random_string 30)
+
+    run_podman run -d --name $rand_name $IMAGE top
+    cid=$output
```
Comment (vrothberg) [inline test/system/085-top.bats:9]: Nit: If we reference the container by ID, we don't need to assign a name.

**Comment by vrothberg** [comment]
>  Do you have any idea what is going on.

No, sorry.  The shlex'ing was added to fix #12452
 but that is a long time ago and the top code is hairy since Podman does things differently.

I _think_ that the compat endpoint should use a different code path than the libpod one and ignore `psgo` entirely.  That may help untangle the code and parsing.

**Comment by vrothberg** [comment]
It may be worth looking at the `docker top` integration tests at this point.

**Comment by Luap99** [comment]
seems like docker doesn't care and just does the same thing Dan does here:
https://github.com/moby/moby/blob/e0da5cb9291982a4c7b3c6b0642eeced4ba9f8ac/client/container_top.go#L16-L18

So overall I would assume if we only get a single ps_args param that we need to `shlex.Split()` which is a bit better than `strings.Split(arg, " ")`. However our remote client should still send an array from the client so when get more than one ps_args arg we can just use the array as is for the ps call. That way we can actually pass down the real args like they were given on the cli.

**Comment by rhatdan** [comment]
Ok I cleaned it up and now I believe I am doing the right thing.  The code in the bindings section could be nicer, suggestions welcome.

**Comment by rhatdan** [comment]
Now back to PTO.

**Comment by TomSweeneyRedHat** [comment]
WHen you get back from PTO, the red tests will be waiting.

File: pkg/api/handlers/compat/containers_top.go [hunk 2]
```diff
@@ -58,7 +59,24 @@ loop: // break out of for/select infinite` loop
...
+				break loop
+			}
+			for _, s := range shSplit {
+				if s != "" {
+					if strings.Contains(s, " ") {
+						s = fmt.Sprintf("%q", s)
+					}
+					psDescriptors = append(psDescriptors, s)
+				}
+			}
```
Comment (Luap99) [inline pkg/api/handlers/compat/containers_top.go:76]: This feels and looks incorrect, the proper fix should be to remove the shlex.Split call in c.Top() in libpod

File: pkg/domain/infra/abi/containers.go [hunk 3]
```diff
@@ -552,7 +553,17 @@ func (ic *ContainerEngine) ContainerTop(ctx context.Context, options entities.To
...
-	report.Value, err = container.Top(options.Descriptors)
+	descriptors := []string{}
+	for _, arg := range options.Descriptors {
+		if strings.Contains(arg, " ") {
+			// Quote multi work args
+			descriptors = append(descriptors, fmt.Sprintf("%q", arg))
+		} else {
+			descriptors = append(descriptors, arg)
+		}
+	}
```
Comment (Luap99) [inline pkg/domain/infra/abi/containers.go:564]: this makes no sense to me as well. On the local path we can juts pass down the array there is no need for this string manipulation.

File: test/e2e/top_test.go [hunk 4]
```diff
@@ -99,8 +99,16 @@ var _ = Describe("Podman top", func() {
 
 		result = podmanTest.Podman([]string{"top", session.OutputToString(), "ax -o args"})
 		result.WaitWithDefaultTimeout()
+		Expect(result).Should(Exit(125))
```
Comment (Luap99) [inline test/e2e/top_test.go:102]: this looks like breaking change to me as the same command used to work fine.

File: pkg/api/handlers/compat/containers_top.go [hunk 5]
```diff
@@ -77,11 +95,7 @@ loop: // break out of for/select infinite` loop
 				}
 
 				for _, line := range output[1:] {
-					process := strings.Split(line, "\t")
-					for i := range process {
-						process[i] = strings.TrimSpace(process[i])
-					}
-					body.Processes = append(body.Processes, process)
+					body.Processes = append(body.Processes, strings.Fields(line))
```
Comment (Luap99) [inline pkg/api/handlers/compat/containers_top.go:98]: why was this done? This changes the output format.

**Review by Luap99 — CHANGES_REQUESTED** [review]
I still think we must use shelex only for a single arg and otherwise use the array as is to provide best compatibility.

File: test/e2e/top_test.go (hunk 4 above)
Comment (rhatdan) [reply test/e2e/top_test.go:102]: This command is invalid in ps, so it should be invalid in podman.  We are correcting the way we pass these commands down to the PS command, before we were always dropping the "" now we pass them down, so this was always broken. Luckily no-one in the real world would use this syntax.

File: pkg/domain/infra/abi/containers.go (hunk 3 above)
Comment (rhatdan) [reply pkg/domain/infra/abi/containers.go:564]: This issue is passing -o "foo bar" is valid for PS.  We need to maintain the "" in this call to pass it down to ps command.

File: pkg/api/handlers/compat/containers_top.go (hunk 2 above)
Comment (rhatdan) [reply pkg/api/handlers/compat/containers_top.go:76]: Currently there are lots of use cases that are broken in the local and remote that are valid for ps command.  The code changes are all about getting the commands correct that we pass to the ps command.

File: pkg/api/handlers/compat/containers_top.go (hunk 5 above)
Comment (rhatdan) [reply pkg/api/handlers/compat/containers_top.go:98]: The output format can not be made to match between podman --remote top and podman top, because of the handling of the API.

This change was made to make this least different, but it ain't pretty.

We could change --remote=false to be the same way, but that would ruin its decent display.

**Comment by rhatdan** [comment]
I can remove the shelex from the c.Top altogether and force both callers to do it themselves.

**Comment by vrothberg** [comment]
> I can remove the shelex from the c.Top altogether and force both callers to do it themselves.

As mentioned above, I think that the compat endpoint should not attempt on using `psgo` at all and call `ps` directly.  Wouldn't that address the issue?

**Comment by rhatdan** [comment]
We can dumb down the Compat endpoint if you want, but that does not solve the problem.  The current code handles all of the test cases, if you have other test cases that fail, please present them and I will fix them.  The only "breakage" I see here was a case where we were passing a invalid command to ps, and parsing it out to a valid command, while we were breaking valid commands.

```
ps "ax -o args"
error: unsupported option (BSD syntax)

Usage:
 ps [options]

 Try 'ps --help <simple|list|output|threads|misc|all>'
  or 'ps --help <s|l|o|t|m|a>'
 for additional help text.

For more details see ps(1).

$ ps "ax -o args"
error: unsupported option (BSD syntax)

Usage:
 ps [options]

 Try 'ps --help <simple|list|output|threads|misc|all>'
  or 'ps --help <s|l|o|t|m|a>'
 for additional help text.

For more details see ps(1).

$ ps ax -o "pid args"
    PID COMMAND
      1 /usr/lib/systemd/systemd --switched-root --system --deserialize=35 rhgb
      2 [kthreadd]
```
Previous podman top succeeded and failed opposite of the ps command.
```
 $ podman top dan "ax -o args"
COMMAND
sleep inf
bluechi (rpm) $ podman top dan ax -o "pid, args"
Error: ps(1) failed with exit code 1: error: conflicting format options

Usage:
 ps [options]

 Try 'ps --help <simple|list|output|threads|misc|all>'
  or 'ps --help <s|l|o|t|m|a>'
 for additional help text.

For more details see ps(1).

 $ podman top dan "ax -o args"
COMMAND
sleep inf
```
This PR Fixes the local client and the remote client.

**Comment by rhatdan** [comment]
@baude @Luap99 @vrothberg @giuseppe Can we get this in and then look at removing functionality from compatibility endpoint, if that is desired.

**Review by vrothberg — COMMENTED** [review]
I will do some more testing.  I find it very hard to follow what this PR does as some many things seem to change.

An older podman client should be able to talk to a newer client; I will check whether that will work.

**Comment by vrothberg** [comment]
`ax -o pid,args` does not work across client/server versions anymore

**Comment by rhatdan** [comment]
Works for me.

```
$ ./bin/podman --remote top 2772a9463cab ax -o pid,args
PID         COMMAND
1           sleep       infinity
```

**Comment by vrothberg** [comment]
> Works for me.

Try it where either client or server are not from this commit.

**Comment by vrothberg** [comment]
> Don't see how we can fix the problems, without changing the client and server.

Split the compat from the libpod code.  If the libpod endpoint needs a new format for the parameter, we can add a new parameter.

**Comment by rhatdan** [comment]
Your just confusing callers then and going to cause long term dept, for broken interfaces.

**Comment by vrothberg** [comment]
> Your just confusing callers then and going to cause long term dept, for broken interfaces.

That's SWE best practices.  The other alternative to break existing users. I picked the example from #12452 which was fixed by adding shlex-ing.  With this PR, we would regress exactly on that.

**Comment by rhatdan** [comment]
In this case I would argue that Docker is broken since it accepts broken syntax and does not accept correct syntax from a ps sense.

"ax -o pid,args" works with docker, which does not with ps
ax -o "pid args" does not work with docker, which does work with ps.

**Comment by rhatdan** [comment]
If I leave the old format for Docker, it should continue to work.  and support older clients, and then newer clients should fix the issue.

File: pkg/api/handlers/compat/containers_top.go (hunk 5 above)
Comment (rhatdan) [reply pkg/api/handlers/compat/containers_top.go:98]: Yes I know, but there is not a better format I can figure out. Leaving the existing format looks aweful also.

**Comment by Luap99** [comment]
Compatibility between different client and server version is important, this is basically the default for podman machines as we do not control client and server machine.

I still think the best way is to treat a single argument as we have to use shelx and if we got more than one we use the array as it. 

I was thinking something like this:

```diff
diff --git a/libpod/container_top_linux.go b/libpod/container_top_linux.go
index 0f3e92a54..41002eb52 100644
--- a/libpod/container_top_linux.go
+++ b/libpod/container_top_linux.go
@@ -214,10 +214,11 @@ func (c *Container) Top(descriptors []string) ([]string, error) {
                return nil, psgoErr
        }
 
-       // Note that the descriptors to ps(1) must be shlexed (see #12452).
-       psDescriptors := []string{}
-       for _, d := range descriptors {
-               shSplit, err := shlex.Split(d)
+       psDescriptors := descriptors
+       if len(descriptors) == 1 {
+               // Note that the descriptors to ps(1) must be shlexed (see #12452).
+               psDescriptors = make([]string, 0, len(descriptors))
+               shSplit, err := shlex.Split(descriptors[0])
                if err != nil {
                        return nil, fmt.Errorf("parsing ps args: %w", err)
                }
diff --git a/pkg/api/handlers/compat/containers_top.go b/pkg/api/handlers/compat/containers_top.go
index b6edd51b4..81eee6760 100644
--- a/pkg/api/handlers/compat/containers_top.go
+++ b/pkg/api/handlers/compat/containers_top.go
@@ -18,14 +18,14 @@ func TopContainer(w http.ResponseWriter, r *http.Request) {
        runtime := r.Context().Value(api.RuntimeKey).(*libpod.Runtime)
        decoder := utils.GetDecoder(r)
 
-       psArgs := "-ef"
-       if utils.IsLibpodRequest(r) {
-               psArgs = ""
+       var psArgs []string
+       if !utils.IsLibpodRequest(r) {
+               psArgs = []string{"-ef"}
        }
        query := struct {
-               Delay  int    `schema:"delay"`
-               PsArgs string `schema:"ps_args"`
-               Stream bool   `schema:"stream"`
+               Delay  int      `schema:"delay"`
+               PsArgs []string `schema:"ps_args"`
+               Stream bool     `schema:"stream"`
        }{
                Delay:  5,
                PsArgs: psArgs,
@@ -52,13 +52,19 @@ func TopContainer(w http.ResponseWriter, r *http.Request) {
 
        encoder := json.NewEncoder(w)
 
+       args := query.PsArgs
+       if len(args) == 1 {
+               // Ugly workaround for older clients which used to send arguments comma separated.
+               args = strings.Split(args[0], ",")
+       }
+
 loop: // break out of for/select infinite` loop
        for {
                select {
                case <-r.Context().Done():
                        break loop
                default:
-                       output, err := c.Top(strings.Split(query.PsArgs, ","))
+                       output, err := c.Top(args)
                        if err != nil {
                                if !statusWritten {
                                        utils.InternalServerError(w, err)
diff --git a/pkg/api/server/register_containers.go b/pkg/api/server/register_containers.go
index 630da119c..d0f15e747 100644
--- a/pkg/api/server/register_containers.go
+++ b/pkg/api/server/register_containers.go
@@ -1173,8 +1173,9 @@ func (s *APIServer) registerContainersHandlers(r *mux.Router) error {
        //    default: 5
        //  - in: query
        //    name: ps_args
-       //    type: string
-       //    default:
+       //    type: array
+       //    items:
+       //       type: string
        //    description: |
        //      arguments to pass to ps such as aux.
        // produces:
diff --git a/pkg/bindings/containers/containers.go b/pkg/bindings/containers/containers.go
index e776a6092..0ca579a6c 100644
--- a/pkg/bindings/containers/containers.go
+++ b/pkg/bindings/containers/containers.go
@@ -290,8 +290,10 @@ func Top(ctx context.Context, nameOrID string, options *TopOptions) ([]string, e
        }
        params := url.Values{}
        if options.Changed("Descriptors") {
-               psArgs := strings.Join(options.GetDescriptors(), ",")
-               params.Add("ps_args", psArgs)
+               psArgs := options.GetDescriptors()
+               for _, arg := range psArgs {
+                       params.Add("ps_args", arg)
+               }
        }
        response, err := conn.DoRequest(ctx, nil, http.MethodGet, "/containers/%s/top", params, nil, nameOrID)
        if err != nil {
```
This doesn't address the case new client old server so we would need some version check on client for better compatibility. And likely the split on comma isn't right either as it breaks some arguments.

Also this does not even touch the total mess of output which is send on the api. It is already broken but I don't think this PR makes it better.

**Comment by rhatdan** [comment]
@Luap99 PTAL

File: test/system/085-top.bats [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    is "${lines[0]}" "USER[ \t]*PID[ \t]*PPID[ \t]*%CPU[ \t]*ELAPSED[ \t]*TTY[ \t]*TIME[ \t]*COMMAND" "podman top"
+    is "${lines[1]}" "root[ \t]*1[ \t]*0[ \t]*0.000[ \t]*" "podman top"
+
+    run_podman top $cid -eo pid,comm
+    is "${lines[0]}" "[ \t]*PID[ \t]*COMMAND" "podman top -eo pid,comm Heading"
+    is "${lines[1]}" "[ \t]*1[ \t]*top" "podman top -eo pid,comm processes"
+
+    run_podman top $cid -eo "pid comm"
+    is "${lines[0]}" "[ \t]*PID[ \t]*COMMAND" "podman top -eo pid,comm Heading"
+    is "${lines[1]}" "[ \t]*1[ \t]*top" "podman top -eo pid,comm processes"
```
Comment (edsantiago) [inline test/system/085-top.bats:20]: copypasta error: the description should be space, not comma ("pid comm")

File: pkg/api/handlers/compat/containers_top.go [hunk 7]
```diff
@@ -52,13 +52,19 @@ func TopContainer(w http.ResponseWriter, r *http.Request) {
 
 	encoder := json.NewEncoder(w)
 
+	args := query.PsArgs
+	if len(args) == 1 {
```
Comment (Luap99) [inline pkg/api/handlers/compat/containers_top.go:56]: I think we should also guard this with `utils.IsLibpodRequest(r)` and `utils.SupportedVersion(r, "< 4.8.0")`, then we only do this for the old libpod API.

The docker API will only ever send us a single arg which will already be shlex.Split() in libpod backend so this should be fine.

**Review by Luap99 — COMMENTED** [review]
The problem is that this still does not fully fix the problem with different API versions for server and client.

And new client will send an array while an old server would only use a single arg.
Because new client old server is common for podman machine this is undesirable.
This needs actual version negation to address that.

File: test/system/085-top.bats [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+load helpers
+
+@test "podman top - basic tests" {
+    run_podman run -d $IMAGE top
+    cid=$output
+    is "$cid" "[0-9a-f]\{64\}$"
+
+    run_podman top $cid
+    is "${lines[0]}" "USER[ \t]*PID[ \t]*PPID[ \t]*%CPU[ \t]*ELAPSED[ \t]*TTY[ \t]*TIME[ \t]*COMMAND" "podman top"
+    is "${lines[1]}" "root[ \t]*1[ \t]*0[ \t]*0.000[ \t]*" "podman top"
```
Comment (edsantiago) [inline test/system/085-top.bats:12]: I'm suspecting that this `0.000` is going to cause flakes, but we shall see...
